### PR TITLE
fix: fail `init --forcePush` fast if environment parameters or secrets are missing in the environment

### PR DESCRIPTION
**Pull request by edwardfoyle**
#### Description of changes
Adds the same check that is run on push [here](https://github.com/aws-amplify/amplify-cli/blob/33f9b3415ce73d5318bb065f3e970daa90b8b8f1/packages/amplify-cli/src/extensions/amplify-helpers/push-resources.ts#L125-L135) to also be run during `init --forcePush`. The raisePrePushEvent function is now called during init if the `--forcePush` flag is set. This allows the function category to run it's pre-push checks but these checks need to run after the project has initialized (hence why it cannot register for a preInit event)

A couple error messages have also been improved with better messaging and resolution steps.

#### Description of how you validated changes
Manually tested, added unit tests and running e2e suite here: https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli?branch=run-e2e/foyleef/fail-fast-checks

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by edwardfoyle** [comment]
> Do we have at least one e2e test that would call `amplify init --forcePush` ?

It doesn't look like we do. I can add one.

File: packages/amplify-e2e-core/src/init/non-interactive-init.ts [hunk 1]
```diff
@@ -48,6 +48,7 @@ export const nonInteractiveInitWithForcePushAttach = async (
       awscloudformation: awsProviderConfig,
     }),
     '--forcePush',
+    '--debug',
```
Comment (sobolk) [inline packages/amplify-e2e-core/src/init/non-interactive-init.ts:51]: should this debug stay ?

**Review by sobolk — COMMENTED** [review]
As for the e2e test.
It worries me that the test we have is a migration test. I.e. when we drop support V10 we'll end up loosing this coverage. I'd rather add some non-migration test for forcePush.

File: packages/amplify-category-function/src/events/postEnvRemoveHandler.ts [hunk 2]
```diff
@@ -1,8 +1,18 @@
...
 export const postEnvRemoveHandler = async (context: $TSContext, envName: string) => {
-  await removeAllEnvSecrets(context, envName);
+  try {
+    await removeAllEnvSecrets(context, envName);
+  } catch (err) {
+    // catching this silently because that was the previous behavior
+    // previously this error was caught by the platform event firing logic and silently ignored
+    // now we are ignoring errors at the individual event handler
+    printer.debug(`function category postEnvRemoveHandler failed to run.`);
+    printer.debug(`You may need to manually clean up some function state.`);
```
Comment (akshbhu) [inline packages/amplify-category-function/src/events/postEnvRemoveHandler.ts:13]: Do we need to throw a amplify Error with resolution or doc link on how to cleanup ?

**Review by akshbhu — DISMISSED** [review]

File: packages/amplify-e2e-core/src/init/non-interactive-init.ts (hunk 1 above)
Comment (edwardfoyle) [reply packages/amplify-e2e-core/src/init/non-interactive-init.ts:51]: I think it can. We run other e2e commands with the flag to help with debugging issues more quickly

File: packages/amplify-category-function/src/events/postEnvRemoveHandler.ts (hunk 2 above)
Comment (edwardfoyle) [reply packages/amplify-category-function/src/events/postEnvRemoveHandler.ts:13]: This case was previously ignored, so this matches the existing behavior but with some additional debug lines

**Comment by edwardfoyle** [comment]
Green e2e run here: https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli/17886/workflows/b76c66bd-1c8b-4739-a96d-572df102889f
Only the last 3 commits are new since then (fixing lint errors and pulling out some of the e2e test harness updates into their own PR)

File: packages/amplify-e2e-tests/src/__tests__/init-force-push.test.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  getTeamProviderInfo,
+  gitCleanFdx,
+  gitCommitAll,
+  gitInit,
+  initJSProjectWithProfile,
+  nonInteractiveInitWithForcePushAttach,
+  setTeamProviderInfo,
+} from '@aws-amplify/amplify-e2e-core';
+
+describe('init --forcePush', () => {
```
Comment (sobolk) [inline packages/amplify-e2e-tests/src/__tests__/init-force-push.test.ts:21]: thank you!

**Review by sobolk — APPROVED** [review]

File: packages/amplify-cli/src/amplify-exception-handler.ts [hunk 4]
```diff
@@ -18,6 +18,7 @@ export const init = (_context: Context): void => {
  * Handle exceptions
  */
 export const handleException = async (exception: unknown): Promise<void> => {
+  process.exitCode = 1;
```
Comment (lazpavel) [inline packages/amplify-cli/src/amplify-exception-handler.ts:21]: why is this needed here?

Comment (akshbhu) [inline packages/amplify-cli/src/amplify-exception-handler.ts:21]: why is this needed when we have a call at the end of this function ?

Comment (edwardfoyle) [reply packages/amplify-cli/src/amplify-exception-handler.ts:21]: I observed a case where if the exception handler errored out before it got to the `process.exit(1)` it would exit with 0 even though there was a failure. This ensures that if we enter the exception handler, we will exit with 1 regardless of what happens
